x/leverage/types/leverage.go

```diff
@@ -0,0 +1,137 @@
+	maxRepay := availableRepay.ToDec()
+	// Determine the base maxReward amount that would result from maximum repayment
+	maxReward := maxRepay.Mul(repayTokenPrice).Mul(sdk.OneDec().Add(liquidationIncentive)).Quo(rewardTokenPrice)
+	// Determine the maxCollateral burn amount that corresponds to base reward amount
+	maxCollateral := maxReward.Quo(uTokenExchangeRate)
+
+	// Catch no-ops early
+	if maxRepay.IsZero() || maxReward.IsZero() || maxCollateral.IsZero() || closeFactor.IsZero() || borrowedValue.IsZero() {
+		return sdk.ZeroInt(), sdk.ZeroInt(), sdk.ZeroInt()
+	}
+
+	// We will track limiting factors by the ratio by which the max repayment would need to be reduced to comply
+	ratio := sdk.OneDec()
+	// Repaid value cannot exceed borrowed value times close factor
+	ratio = sdk.MinDec(ratio,
+		borrowedValue.Mul(closeFactor).Quo(maxRepay.Mul(repayTokenPrice)),
+	)
+	// Collateral burned cannot exceed borrower's collateral
+	ratio = sdk.MinDec(ratio,
+		availableCollateral.ToDec().Quo(maxCollateral),
+	)
+	// Base token reward cannot exceed available unreserved module balance
+	ratio = sdk.MinDec(ratio,
+		availableReward.ToDec().Quo(maxReward),
+	)
+	// Catch edge cases
+	ratio = sdk.MaxDec(ratio, sdk.ZeroDec())
+
+	// Reduce repay and collateral limits by the most severe limiting factor encountered
+	maxRepay = maxRepay.Mul(ratio)
+	maxCollateral = maxCollateral.Mul(ratio)
+
+	// No rounding has occurred yet, but both values are now within the
+	// limits defined by available balances and module parameters.
+
+	// First, the amount of borrowed token the liquidator must repay is rounded up.
+	// This is a slight disadvantage to the liquidator in favor of the borrower and
+	// the module. It also ensures borrow dust is always eliminated when encountered.
+	tokenRepay = maxRepay.Ceil().RoundInt()
+
+	// Next, the amount of collateral uToken the borrower will lose is rounded down.
+	// This is favors the borrower over the liquidator, and also protects the module.
+	collateralBurn = maxCollateral.TruncateInt()
+
+	// One danger to rounding collateral burn down is that of collateral dust. This
+	// can be considered in two scenarios:
+	// 1) If collateral was the limiting factor above, then it will have already been
+	// an integer amount and truncating is a no-op.
+	// 2) If collateral was not the limiting factor, then there will be a non-dust
+	// quantity left over anyway.
+
+	// Finally, the base token reward amount is derived directly from the collateral
+	// to burn. This will round down identically to MsgWithdraw, favoring the module
+	// over the liquidator.
+	tokenReward = collateralBurn.ToDec().Mul(uTokenExchangeRate).TruncateInt()
+
+	return tokenRepay, collateralBurn, tokenReward
+}
+
+// ComputeCloseFactor uses a borrower's borrowed value and liquidation threshold and
+// some leverage module parameters to derive a dynamic close factor for liquidation.
+func ComputeCloseFactor(
```

Comment:
Same as above - this function (formerly called `LiquidationParams`) was moved from `keeper` to `types`.

Also, it now computes only `CloseFactor`, instead of also passing through `LiquidationIncentive` unmodified.

Reply:
Moved back to original location - functionality improvement is retained.